File: sdk/python/src/pelisearch/documents.py
from __future__ import annotations

from typing import Any, Callable

from .types import BulkDocumentResult, BulkResponse, DocumentCreatedResponse


class DocumentOperations:
    def __init__(self, request: Callable[..., object]) -> None:
        self._request = request

    def add_document(
        self, index: str, id: str, fields: dict[str, Any]
    ) -> DocumentCreatedResponse:
        body = self._request(
            "POST",
            f"/indexes/{_esc(index)}/documents",
            {"id": id, "fields": fields},
        )
        return DocumentCreatedResponse(**body)

    def get_document(self, index: str, id: str) -> dict[str, Any]:
        return self._request(
            "GET",
            f"/indexes/{_esc(index)}/documents/{_esc(id)}",
        )

    def delete_document(self, index: str, id: str) -> None:
        self._request(
            "DELETE",
            f"/indexes/{_esc(index)}/documents/{_esc(id)}",
        )

    def bulk_add_documents(
        self,
        index: str,
        documents: list[dict[str, Any]],
    ) -> BulkResponse:
        body = self._request(
            "POST",
            f"/indexes/{_esc(index)}/documents/bulk",
            {"documents": documents},
        )
        docs = [BulkDocumentResult(**d) for d in body["documents"]]
        return BulkResponse(documents=docs)


def _esc(s: str) -> str:
    import urllib.parse

    return urllib.parse.quote(s, safe="")

File: sdk/javascript/src/documents.ts
import type { RequestFn, DocumentCreatedResponse, BulkResponse } from "./types.js"

export interface AddDocumentPayload {
  id: string
  fields: Record<string, unknown>
}

export class DocumentsModule {
  constructor(private readonly request: RequestFn) {}

  async add(
    index: string,
    id: string,
    fields: Record<string, unknown>,
  ): Promise<DocumentCreatedResponse> {
    return this.request<DocumentCreatedResponse>(
      "POST",
      `/indexes/${encodeURIComponent(index)}/documents`,
      { id, fields },
    )
  }

  async get(index: string, id: string): Promise<Record<string, unknown>> {
    return this.request<Record<string, unknown>>(
      "GET",
      `/indexes/${encodeURIComponent(index)}/documents/${encodeURIComponent(id)}`,
    )
  }

  async delete(index: string, id: string): Promise<void> {
    await this.request<void>(
      "DELETE",
      `/indexes/${encodeURIComponent(index)}/documents/${encodeURIComponent(id)}`,
    )
  }

  async bulkAdd(
    index: string,
    documents: AddDocumentPayload[],
  ): Promise<BulkResponse> {
    return this.request<BulkResponse>(
      "POST",
      `/indexes/${encodeURIComponent(index)}/documents/bulk`,
      { documents },
    )
  }
}
